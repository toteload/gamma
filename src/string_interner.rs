use serde::Serialize;
use std::collections::HashMap;

#[derive(PartialEq, Debug, Clone, Copy, Hash, Eq, PartialOrd, Ord, Serialize)]
pub struct Symbol(u32);

#[derive(Serialize)]
pub struct StringInterner {
    symbols: HashMap<Box<str>, Symbol>,
    strings: Vec<Box<str>>,
}

// This implementation also looks good and simple.
// https://github.com/rust-lang/rust/blob/89b9f7b284aacc5f8613438b80e4dd7bdd10549e/compiler/rustc_span/src/symbol.rs#L1741

impl StringInterner {
    pub fn new() -> StringInterner {
        StringInterner {
            symbols: HashMap::new(),
            strings: Vec::new(),
        }
    }

    pub fn add(&mut self, name: &str) -> Symbol {
        if let Some(sym) = self.symbols.get(name) {
            return *sym;
        }

        let sym = Symbol(self.strings.len() as u32);

        let s = name.to_string().into_boxed_str();

        self.strings.push(s.clone());

        self.symbols.insert(s, sym);

        sym
    }

    pub fn find_symbol(&self, name: &str) -> Option<Symbol> {
        self.symbols.get(name).copied()
    }

    pub fn get(&self, symbol: &Symbol) -> &str {
        &self.strings[symbol.0 as usize]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basic_string_interner_usage() {
        let mut interner = StringInterner::new();

        let a = interner.add("banana bread");
        let b = interner.add("chocolate");
        let c = interner.add("tuna");
        let d = interner.add("tuna");

        assert_eq!(c, d);

        assert_eq!(interner.get(&a), "banana bread");
        assert_eq!(interner.get(&b), "chocolate");
        assert_eq!(interner.get(&c), "tuna");

        let d = interner.add("chocolate");

        assert_eq!(b, d);

        assert_eq!(interner.find_symbol("tuna").unwrap(), interner.add("tuna"));
    }
}
