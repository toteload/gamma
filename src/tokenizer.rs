use crate::source_location::{SourceLocation, SourceSpan};
use crate::string_interner::{StringInterner, Symbol};
use serde::Serialize;
use std::iter::Peekable;
use std::str::Chars;

#[derive(Clone, Copy, Debug, Serialize)]
pub struct Token {
    pub span: SourceSpan,
    pub kind: TokenKind,
}

#[derive(strum_macros::IntoStaticStr, Clone, Copy, Debug, Hash, PartialEq, Eq, Serialize)]
pub enum TokenKind {
    KeywordFn,

    KeywordIf,
    KeywordElse,
    KeywordLoop,
    KeywordBreak,
    KeywordContinue,
    KeywordReturn,

    KeywordLet,
    KeywordSet,

    KeywordCast,

    KeywordEnd,

    KeywordOr,
    KeywordAnd,
    KeywordNot,
    KeywordXor,

    KeywordBor,
    KeywordBand,

    // TODO call these keywords?
    Equal,
    NotEqual,
    LessEqual,
    Less,
    GreaterEqual,
    Greater,

    Identifier(Symbol),

    IntLiteral(i64),
    BoolLiteral(bool),

    BraceOpen,
    BraceClose,

    ParenOpen,
    ParenClose,

    Semicolon,
    Colon,

    Comma,

    EqualSign,

    Ampersand,
    At,

    Hat,
    Star,
    Minus,
    Plus,
    Div,
}

pub struct Tokenizer<'a> {
    source: &'a str,
    iter: Peekable<Chars<'a>>,
    loc: SourceLocation,
    offset: usize,
    str_interner: &'a mut StringInterner,
}

impl<'a> Tokenizer<'a> {
    pub fn new(source: &'a str, str_interner: &'a mut StringInterner) -> Self {
        Tokenizer {
            source,
            iter: source.chars().peekable(),
            loc: SourceLocation { line: 1, col: 1 },
            offset: 0,
            str_interner,
        }
    }

    fn advance(&mut self) -> Option<(usize, SourceLocation, char)> {
        let c = self.iter.next()?;

        let offset = self.offset;
        let loc = self.loc;

        if c == '\n' {
            self.loc.line += 1;
            self.loc.col = 1;
        } else {
            self.loc.col += 1;
        }

        self.offset += c.len_utf8();

        Some((offset, loc, c))
    }

    fn skip_whitespace(&mut self) {
        loop {
            let Some(c) = self.iter.peek() else {
                break;
            };

            if !c.is_whitespace() {
                break;
            }

            self.advance();
        }
    }
}

impl<'a> Iterator for Tokenizer<'a> {
    type Item = Token;

    fn next(&mut self) -> Option<Token> {
        self.skip_whitespace();

        let (offset, start, c) = self.advance()?;

        #[rustfmt::skip]
        let tok = match c {
            // Identifier or keyword.
            _ if c.is_ascii_alphabetic() || c == '_' => {
                fn is_identifier_char(c: char) -> bool {
                    c.is_ascii_alphanumeric() || c == '_'
                }

                let (mut end_offset, mut end_loc) = (offset + c.len_utf8(), start);
                loop {
                    let Some(a) = self.iter.peek() else { break; };

                    if !is_identifier_char(*a) {
                        break;
                    }

                    let b;
                    (_, end_loc, b) = self.advance().unwrap();
                    end_offset += b.len_utf8();
                }

                let identifier = &self.source[offset..end_offset];

                let span = SourceSpan {
                    start,
                    end: end_loc,
                };

                match identifier {
                    "fn"       => Token { span, kind: TokenKind::KeywordFn, },
                    "if"       => Token { span, kind: TokenKind::KeywordIf, },
                    "else"     => Token { span, kind: TokenKind::KeywordElse, },
                    "loop"     => Token { span, kind: TokenKind::KeywordLoop, },
                    "break"    => Token { span, kind: TokenKind::KeywordBreak, },
                    "continue" => Token { span, kind: TokenKind::KeywordContinue, },
                    "return"   => Token { span, kind: TokenKind::KeywordReturn, },
                    "let"      => Token { span, kind: TokenKind::KeywordLet, },
                    "set"      => Token { span, kind: TokenKind::KeywordSet, },
                    "cast"     => Token { span, kind: TokenKind::KeywordCast, },
                    "end"      => Token { span, kind: TokenKind::KeywordEnd, },

                    "and"      => Token { span, kind: TokenKind::KeywordAnd, },
                    "or"       => Token { span, kind: TokenKind::KeywordOr, },
                    "not"      => Token { span, kind: TokenKind::KeywordNot, },
                    "xor"      => Token { span, kind: TokenKind::KeywordXor, },
                    "bor"      => Token { span, kind: TokenKind::KeywordBor, },
                    "band"     => Token { span, kind: TokenKind::KeywordBand, },
                    "eq"       => Token { span, kind: TokenKind::Equal, },
                    "ne"       => Token { span, kind: TokenKind::NotEqual, },
                    "le"       => Token { span, kind: TokenKind::LessEqual, },
                    "lt"       => Token { span, kind: TokenKind::Less, },
                    "ge"       => Token { span, kind: TokenKind::GreaterEqual, },
                    "gt"       => Token { span, kind: TokenKind::Greater, },

                    "false"    => Token { span, kind: TokenKind::BoolLiteral(false), },
                    "true"     => Token { span, kind: TokenKind::BoolLiteral(true), },
                    _ => {
                        let sym = self.str_interner.add(identifier);
                        Token {
                            span,
                            kind: TokenKind::Identifier(sym),
                        }
                    }
                }
            }

            _ if c.is_ascii_digit() => {
                let (mut end_offset, mut end_loc) = (offset + c.len_utf8(), start);
                loop {
                    let Some(d) = self.iter.peek() else { break; };

                    if !d.is_ascii_digit() {
                        break;
                    }

                    let b;
                    (_, end_loc, b) = self.advance().unwrap();
                    end_offset += b.len_utf8();
                }

                let number: i64 = self.source[offset..end_offset].parse().unwrap();

                let span = SourceSpan {
                    start,
                    end: end_loc,
                };

                Token {
                    span,
                    kind: TokenKind::IntLiteral(number),
                }
            }

            ';' => Token { span: SourceSpan::single(start), kind: TokenKind::Semicolon, },
            ':' => Token { span: SourceSpan::single(start), kind: TokenKind::Colon, },

            '(' => Token { span: SourceSpan::single(start), kind: TokenKind::ParenOpen, },
            ')' => Token { span: SourceSpan::single(start), kind: TokenKind::ParenClose, },
            '{' => Token { span: SourceSpan::single(start), kind: TokenKind::BraceOpen, },
            '}' => Token { span: SourceSpan::single(start), kind: TokenKind::BraceClose, },

            ',' => Token { span: SourceSpan::single(start), kind: TokenKind::Comma, },

            '=' => Token { span: SourceSpan::single(start), kind: TokenKind::EqualSign, },

            '*' => Token { span: SourceSpan::single(start), kind: TokenKind::Star, },
            '+' => Token { span: SourceSpan::single(start), kind: TokenKind::Plus, },
            '/' => Token { span: SourceSpan::single(start), kind: TokenKind::Div, },
            '-' => Token { span: SourceSpan::single(start), kind: TokenKind::Minus, },
            '^' => Token { span: SourceSpan::single(start), kind: TokenKind::Hat, },
            '&' => Token { span: SourceSpan::single(start), kind: TokenKind::Ampersand, },
            '@' => Token { span: SourceSpan::single(start), kind: TokenKind::At, },

            _ => todo!("Character '{}' not recognized", c),
        };

        Some(tok)
    }
}

#[cfg(test)]
mod tests {
    use super::Tokenizer;
    use crate::string_interner::StringInterner;
    use insta::assert_ron_snapshot;

    #[test]
    fn tokenizer_creates_the_expected_tokens() {
        let source = "fn if else + - * ; : { } ( ) void let voidlet 4687 continue return-,=       
            int eq / ne ge le lt gt true false loop break bool int void end";

        let mut symbols = StringInterner::new();
        let tokenizer = Tokenizer::new(source, &mut symbols);

        let tokens = tokenizer.collect::<Vec<_>>();

        assert_ron_snapshot!("tokens_snapshot", tokens);
    }
}
