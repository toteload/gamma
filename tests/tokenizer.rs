use gamma::string_interner::StringInterner;
use gamma::tokenizer::Tokenizer;
use insta::assert_ron_snapshot;

#[test]
fn tokenizer_creates_the_expected_tokens() {
    let source = "fn if else + - * ; : { } ( ) void let voidlet 4687 continue return->,=       
int &&& == ^ & | || !! / != >= <= < > true false loop break bool int void ";

    let mut symbols = StringInterner::new();
    let tokenizer = Tokenizer::new(source, &mut symbols);

    let tokens = tokenizer.collect::<Vec<_>>();

    assert_ron_snapshot!("tokens_snapshot", tokens);
}
