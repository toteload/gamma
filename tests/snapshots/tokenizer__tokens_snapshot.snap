---
source: tests/tokenizer.rs
expression: tokens
---
[
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 1,
      ),
      end: SourceLocation(
        line: 1,
        col: 2,
      ),
    ),
    kind: KeywordFn,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 4,
      ),
      end: SourceLocation(
        line: 1,
        col: 5,
      ),
    ),
    kind: KeywordIf,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 7,
      ),
      end: SourceLocation(
        line: 1,
        col: 10,
      ),
    ),
    kind: KeywordElse,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 12,
      ),
      end: SourceLocation(
        line: 1,
        col: 12,
      ),
    ),
    kind: Plus,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 14,
      ),
      end: SourceLocation(
        line: 1,
        col: 14,
      ),
    ),
    kind: Minus,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 16,
      ),
      end: SourceLocation(
        line: 1,
        col: 16,
      ),
    ),
    kind: Star,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 18,
      ),
      end: SourceLocation(
        line: 1,
        col: 18,
      ),
    ),
    kind: Semicolon,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 20,
      ),
      end: SourceLocation(
        line: 1,
        col: 20,
      ),
    ),
    kind: Colon,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 22,
      ),
      end: SourceLocation(
        line: 1,
        col: 22,
      ),
    ),
    kind: BraceOpen,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 24,
      ),
      end: SourceLocation(
        line: 1,
        col: 24,
      ),
    ),
    kind: BraceClose,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 26,
      ),
      end: SourceLocation(
        line: 1,
        col: 26,
      ),
    ),
    kind: ParenOpen,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 28,
      ),
      end: SourceLocation(
        line: 1,
        col: 28,
      ),
    ),
    kind: ParenClose,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 30,
      ),
      end: SourceLocation(
        line: 1,
        col: 33,
      ),
    ),
    kind: KeywordVoid,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 35,
      ),
      end: SourceLocation(
        line: 1,
        col: 37,
      ),
    ),
    kind: KeywordLet,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 39,
      ),
      end: SourceLocation(
        line: 1,
        col: 45,
      ),
    ),
    kind: Identifier(Symbol(0)),
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 47,
      ),
      end: SourceLocation(
        line: 1,
        col: 50,
      ),
    ),
    kind: IntLiteral(4687),
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 52,
      ),
      end: SourceLocation(
        line: 1,
        col: 59,
      ),
    ),
    kind: KeywordContinue,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 61,
      ),
      end: SourceLocation(
        line: 1,
        col: 66,
      ),
    ),
    kind: KeywordReturn,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 67,
      ),
      end: SourceLocation(
        line: 1,
        col: 68,
      ),
    ),
    kind: Arrow,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 69,
      ),
      end: SourceLocation(
        line: 1,
        col: 69,
      ),
    ),
    kind: Comma,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 1,
        col: 70,
      ),
      end: SourceLocation(
        line: 1,
        col: 70,
      ),
    ),
    kind: Equals,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 1,
      ),
      end: SourceLocation(
        line: 2,
        col: 3,
      ),
    ),
    kind: KeywordInt,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 5,
      ),
      end: SourceLocation(
        line: 2,
        col: 6,
      ),
    ),
    kind: LogicalAnd,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 7,
      ),
      end: SourceLocation(
        line: 2,
        col: 7,
      ),
    ),
    kind: BitwiseAnd,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 9,
      ),
      end: SourceLocation(
        line: 2,
        col: 10,
      ),
    ),
    kind: CmpEq,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 12,
      ),
      end: SourceLocation(
        line: 2,
        col: 12,
      ),
    ),
    kind: Xor,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 14,
      ),
      end: SourceLocation(
        line: 2,
        col: 14,
      ),
    ),
    kind: BitwiseAnd,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 16,
      ),
      end: SourceLocation(
        line: 2,
        col: 16,
      ),
    ),
    kind: BitwiseOr,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 18,
      ),
      end: SourceLocation(
        line: 2,
        col: 19,
      ),
    ),
    kind: LogicalOr,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 21,
      ),
      end: SourceLocation(
        line: 2,
        col: 21,
      ),
    ),
    kind: Not,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 22,
      ),
      end: SourceLocation(
        line: 2,
        col: 22,
      ),
    ),
    kind: Not,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 24,
      ),
      end: SourceLocation(
        line: 2,
        col: 24,
      ),
    ),
    kind: Div,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 26,
      ),
      end: SourceLocation(
        line: 2,
        col: 27,
      ),
    ),
    kind: CmpNe,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 29,
      ),
      end: SourceLocation(
        line: 2,
        col: 30,
      ),
    ),
    kind: CmpGe,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 32,
      ),
      end: SourceLocation(
        line: 2,
        col: 33,
      ),
    ),
    kind: CmpLe,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 35,
      ),
      end: SourceLocation(
        line: 2,
        col: 35,
      ),
    ),
    kind: CmpLt,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 37,
      ),
      end: SourceLocation(
        line: 2,
        col: 37,
      ),
    ),
    kind: CmpGt,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 39,
      ),
      end: SourceLocation(
        line: 2,
        col: 42,
      ),
    ),
    kind: BoolLiteral(true),
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 44,
      ),
      end: SourceLocation(
        line: 2,
        col: 48,
      ),
    ),
    kind: BoolLiteral(false),
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 50,
      ),
      end: SourceLocation(
        line: 2,
        col: 53,
      ),
    ),
    kind: KeywordLoop,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 55,
      ),
      end: SourceLocation(
        line: 2,
        col: 59,
      ),
    ),
    kind: KeywordBreak,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 61,
      ),
      end: SourceLocation(
        line: 2,
        col: 64,
      ),
    ),
    kind: KeywordBool,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 66,
      ),
      end: SourceLocation(
        line: 2,
        col: 68,
      ),
    ),
    kind: KeywordInt,
  ),
  Token(
    span: SourceSpan(
      start: SourceLocation(
        line: 2,
        col: 70,
      ),
      end: SourceLocation(
        line: 2,
        col: 73,
      ),
    ),
    kind: KeywordVoid,
  ),
]
